lib/Sema/CSSimplify.cpp

```diff
@@ -1213,6 +1234,68 @@ class ArgumentFailureTracker : public MatchCallArgumentListener {
   }
 };
 
+/// Ignores any failures after the code comletion token.
+class CompletionArgumentTracker : public ArgumentFailureTracker {
+  struct CompletionArgInfo ArgInfo;
+
+public:
+  CompletionArgumentTracker(ConstraintSystem &cs,
+                            SmallVectorImpl<AnyFunctionType::Param> &args,
+                            ArrayRef<AnyFunctionType::Param> params,
+                            ConstraintLocatorBuilder locator,
+                            struct CompletionArgInfo ArgInfo)
+      : ArgumentFailureTracker(cs, args, params, locator), ArgInfo(ArgInfo) {}
+
+  Optional<unsigned> missingArgument(unsigned paramIdx,
+                                     unsigned argInsertIdx) override {
+    const auto &param = Parameters[paramIdx];
+
+    // When solving for code completion, if any argument contains the
+    // completion location, later arguments shouldn't be considered missing
+    // (causing the solution to have a worse score) as the user just hasn't
+    // written them yet. Early exit to avoid recording them in this case.
+    if (ArgInfo.allowsMissingArgAt(argInsertIdx, param)) {
+      unsigned newArgIdx = Arguments.size();
```

Comment:
I think it's worth extracting this whole body of `if` into a separate method on a `ArgumentFailureTracker`, so it could be used by both versions of `missingArgument`.

Reply:
Good idea 👍